controllers/openstackdataplanerole_controller.go

```diff
@@ -294,6 +305,84 @@ func (r *OpenStackDataPlaneRoleReconciler) ReconcileBaremetalSet(ctx context.Con
 	return nil, nil
 }
 
+// ServiceYAML struct for service YAML unmarshalling
+type ServiceYAML struct {
+	Kind     string
+	Metadata yaml.Node
+	Spec     yaml.Node
+}
+
+// EnsureServices - ensure the OpenStackDataPlaneServices exist
+func (r *OpenStackDataPlaneRoleReconciler) EnsureServices(ctx context.Context, helper *helper.Helper, instance *dataplanev1beta1.OpenStackDataPlaneRole) error {
+	servicesPath, found := os.LookupEnv("OPERATOR_SERVICES")
+	if !found {
+		servicesPath = "config/services"
+		os.Setenv("OPERATOR_SERVICES", servicesPath)
+		util.LogForObject(
+			helper, "OPERATOR_SERVICES not set in env when reconciling ", instance,
+			"defaulting to ", servicesPath)
+	}
+
+	r.Log.Info("Ensuring services", "servicesPath", servicesPath)
+	services, err := os.ReadDir(servicesPath)
+	if err != nil {
+		return err
+	}
+
+	for _, service := range services {
+
+		servicePath := path.Join(servicesPath, service.Name())
+
+		if !strings.HasSuffix(service.Name(), ".yaml") {
+			r.Log.Info("Skipping ensuring service from file without .yaml suffix", "file", service.Name())
+			continue
+		}
+
+		data, _ := os.ReadFile(servicePath)
+		var serviceObj ServiceYAML
+		err = yaml.Unmarshal(data, &serviceObj)
+		if err != nil {
+			r.Log.Info("Service YAML file Unmarshal error", "service YAML file", servicePath)
+			return err
+		}
+
+		if serviceObj.Kind != "OpenStackDataPlaneService" {
+			r.Log.Info("Skipping ensuring service since kind is not OpenStackDataPlaneService", "file", servicePath, "Kind", serviceObj.Kind)
+			continue
+		}
+
+		serviceObjMeta := &metav1.ObjectMeta{}
+		err = serviceObj.Metadata.Decode(serviceObjMeta)
+		if err != nil {
+			r.Log.Info("Service Metadata decode error")
+			return err
+		}
+
+		serviceObjSpec := &dataplanev1beta1.OpenStackDataPlaneServiceSpec{}
+		err = serviceObj.Spec.Decode(serviceObjSpec)
+		if err != nil {
+			r.Log.Info("Service Spec decode error")
+			return err
+		}
+
+		ensureService := &dataplanev1beta1.OpenStackDataPlaneService{
```

Comment:
Are these service definitions CRs immutable? If someone wants to make changes in service definition by changing the file, they would only be patched with reconcile of an old role or a new role is created, right? And how and when these services would be cleaned up?